### PR TITLE
[Keyboard] Update Symmetric70 keyboard prototype

### PR DESCRIPTION
**Pull request by mtei**
## Description

The following additions have been made to the Symmetric70 keyboard prototype.

* Added a version using Proton C.
* Added adaptive delay code.
  | MCU               | matrix.c | delay type    | execution time <br> of <br> `matrix_scan()` | matrix scan rate |
  |-------------------|----------|---------------|------------|----------------|
  | Pro Micro (16MHz) | normal   | default       | 503us      | 1810 scans/sec |
  | Pro Micro (16MHz) | normal   | adaptive      | 383us      | 2320 scans/sec |
  | Pro Micro (16MHz) | normal   | fast adaptive | 382us      | 2330 scans/sec |
  | Proton C (72MHz)  | normal   | default       | 210us      | 4350 scans/sec |
  | Proton C (72MHz)  | normal   | adaptive      | 76.4us     | 10600 scans/sec |
  | Proton C (72MHz)  | normal   | fast adaptive | 75.6us     | 10800 scans/sec |

  see more detail "[normal matrix](https://github.com/mtei/qmk_firmware/tree/update_symmetric70_proto/keyboards/handwired/symmetric70_proto/matrix_debug/readme.md)", "[adaptive delay](https://github.com/mtei/qmk_firmware/blob/7a4ee9e4f4021a1608b655766ea0b234b65270ea/keyboards/handwired/symmetric70_proto/matrix_debug/matrix.c#L173-L185)", "[fast adaptive delay](https://github.com/mtei/qmk_firmware/blob/7a4ee9e4f4021a1608b655766ea0b234b65270ea/keyboards/handwired/symmetric70_proto/matrix_debug/matrix.c#L156-L172)"

* Added fast and extensible matrix.c.
  | MCU               | matrix.c | delay type    | execution time <br> of <br> `matrix_scan()` | matrix scan rate |
  |-------------------|----------|---------------|------------|-----------------|
  | Pro Micro (16MHz) | fast     | fast adaptive | 75.6us     |  8090 scans/sec |
  | Proton C (72MHz)  | fast     | fast adaptive | 49.8us     | 15100 scans/sec |

  see more detail "[fast matrix](https://github.com/mtei/qmk_firmware/tree/update_symmetric70_proto/keyboards/handwired/symmetric70_proto/matrix_fast/readme.md)"

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/handwired/symmetric70_proto/local_features.mk [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+#    make MTEST=scan symmetric70_proto/{promicro|proton_c}/{fast|normal}:default:flash
+#  set MATRIX_DEBUG_DELAY and MATRIX_IO_DELAY macro
+#    make MTEST=matrix_debug_delay,mdelay0 symmetric70_proto/{promicro|proton_c}/{fast|normal}:default:flash
+#  set MATRIX_DEBUG_SCAN
+#    make MTEST=matrix_debug_scan symmetric70_proto/{promicro|proton_c}/{fast|normal}:default:flash
+
+ifneq ($(strip $(MTEST)),)
+  define KEYBOARD_OPTION_PARSE
+    # parse 'consle', 'scan', 'no-scan', 'mdelay=?', 'mdelay0',
+    #       'adaptive_delay', 'allways_delay', 'matrix_debug_delay', 'matrix_debug_scan'
```
Comment (tzarc) [inline keyboards/handwired/symmetric70_proto/local_features.mk:16]:
```suggestion
    #       'adaptive_delay', 'always_delay', 'matrix_debug_delay', 'matrix_debug_scan'
```

File: keyboards/handwired/symmetric70_proto/local_features.mk [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        ADAPTIVE_DELAY = yes
+    endif
+    ifeq ($(strip $1),adaptive_delay2)
+        ADAPTIVE_DELAY2 = yes
+    endif
+    ifeq ($(strip $1),adaptive_delay_fast)
+        ADAPTIVE_DELAY_FAST = yes
+    endif
+    ifeq ($(strip $1),allways_delay)
+        ALLWAYS_DELAY = yes
```
Comment (tzarc) [inline keyboards/handwired/symmetric70_proto/local_features.mk:46]:
```suggestion
    ifeq ($(strip $1),always_delay)
        ALWAYS_DELAY = yes
```

File: keyboards/handwired/symmetric70_proto/local_features.mk [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+ifeq ($(strip $(ADAPTIVE_DELAY2)),yes)
+    OPT_DEFS += -DMATRIX_IO_DELAY_ADAPTIVE2
+endif
+
+ifeq ($(strip $(ADAPTIVE_DELAY_FAST)),yes)
+    OPT_DEFS += -DMATRIX_IO_DELAY_ADAPTIVE_FAST
+endif
+
+ifeq ($(strip $(ALLWAYS_DELAY)),yes)
+    OPT_DEFS += -DMATRIX_IO_DELAY_ALLWAYS
```
Comment (tzarc) [inline keyboards/handwired/symmetric70_proto/local_features.mk:80]:
```suggestion
ifeq ($(strip $(ALWAYS_DELAY)),yes)
    OPT_DEFS += -DMATRIX_IO_DELAY_ALWAYS
```

File: keyboards/handwired/symmetric70_proto/matrix_debug/matrix.c [hunk 4]
```diff
@@ -20,6 +20,24 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+#    define MATRIX_DEBUG_SCAN_END()
+#    define MATRIX_DEBUG_DELAY_START()
+#    define MATRIX_DEBUG_DELAY_END()
+#    define MATRIX_DEBUG_GAP()
+#else
+#    define MATRIX_DEBUG_GAP()  asm volatile("nop \n nop":::"memory")
+#endif
+
+#ifndef MATRIX_IO_DELAY_ALLWAYS
+#    define MATRIX_IO_DELAY_ALLWAYS 0
```
Comment (tzarc) [inline keyboards/handwired/symmetric70_proto/matrix_debug/matrix.c:39]:
```suggestion
#ifndef MATRIX_IO_DELAY_ALWAYS
#    define MATRIX_IO_DELAY_ALWAYS 0
```

File: keyboards/handwired/symmetric70_proto/matrix_debug/matrix.c [hunk 5]
```diff
@@ -122,6 +153,60 @@ static bool read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row)
...
+                writePin(MATRIX_MUL_SELECT,col_sel[col_index]);
+                waitInputPinDelay();
+#endif
+                state |= (readPin(col_pins[col_index]) == 0);
+            }
+            MATRIX_DEBUG_DELAY_END();
+        } while (state);
+    }
+#endif
+    if (MATRIX_IO_DELAY_ALLWAYS || current_row + 1 < MATRIX_ROWS) {
```
Comment (tzarc) [inline keyboards/handwired/symmetric70_proto/matrix_debug/matrix.c:205]:
```suggestion
    if (MATRIX_IO_DELAY_ALWAYS || current_row + 1 < MATRIX_ROWS) {
```

File: keyboards/handwired/symmetric70_proto/matrix_debug/matrix.c [hunk 6]
```diff
@@ -184,6 +266,9 @@ static bool read_rows_on_col(matrix_row_t current_matrix[], uint8_t current_col)
 
     // Unselect col
     unselect_col(current_col);
+    if (MATRIX_IO_DELAY_ALLWAYS || current_col + 1 < MATRIX_COLS) {
```
Comment (tzarc) [inline keyboards/handwired/symmetric70_proto/matrix_debug/matrix.c:269]:
```suggestion
    if (MATRIX_IO_DELAY_ALWAYS || current_col + 1 < MATRIX_COLS) {
```

File: keyboards/handwired/symmetric70_proto/matrix_debug/readme.md [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+  * `make MTEST=mdelay=5 handwired/symmetric70_proto/promicro/normal:default:flash`
+  * `make MTEST=mdelay=10 handwired/symmetric70_proto/promicro/normal:default:flash`
+  * `make MTEST=mdelay=20 handwired/symmetric70_proto/promicro/normal:default:flash`
+  * `make MTEST=mdelay=30 handwired/symmetric70_proto/promicro/normal:default:flash`
+* Measure the execution time of matrix_scan()
+  * `make MTEST=matrix_debug_scan[,<other options>..] handwired/symmetric70_proto/promicro/normal:default:flash`
+* Measure delay time.
+  * `make MTEST=matrix_debug_delay[,<other options>..] handwired/symmetric70_proto/promicro/normal:default:flash`
+* Change the behavior of delay
+  * `make MTEST=matrix_debug_delay,allways_delay handwired/symmetric70_proto/promicro/normal:default:flash`
```
Comment (tzarc) [inline keyboards/handwired/symmetric70_proto/matrix_debug/readme.md:26]:
```suggestion
  * `make MTEST=matrix_debug_delay,always_delay handwired/symmetric70_proto/promicro/normal:default:flash`
```

File: keyboards/handwired/symmetric70_proto/matrix_debug/readme.md [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+ - CH1: Row 0
+ - CH2: Row 1
+ - CH3: Row 4
+ - CH4: matrix_scan()
+ - Execution time of matrix_scan()  503us
+ - Frequency of matrix scan 1.81kHz (551.0us)  
+   ![DS1Z_QuickPrint2](https://user-images.githubusercontent.com/2170248/115994477-0ba64400-a612-11eb-98ba-b8cc362f26ac.png)
+
+#### Allways call `matrix_output_unselect_delay()` (show `matrix_scan()` time, default MATRIX_IO_DELAY)
+ - `make MTEST=matrix_debug_scan,allways_delay handwired/symmetric70_proto/promicro/normal:default:flash`
```
Comment (tzarc) [inline keyboards/handwired/symmetric70_proto/matrix_debug/readme.md:43]:
```suggestion
#### Always call `matrix_output_unselect_delay()` (show `matrix_scan()` time, default MATRIX_IO_DELAY)
 - `make MTEST=matrix_debug_scan,always_delay handwired/symmetric70_proto/promicro/normal:default:flash`
```

File: keyboards/handwired/symmetric70_proto/matrix_debug/readme.md [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+ - CH1: Row 0
+ - CH2: Row 1
+ - CH3: Row 4
+ - CH4: matrix_scan()
+ - Execution time of matrix_scan()  210us
+ - Frequency of matrix scan 4.35kHz (230.0us)  
+   ![DS1Z_QuickPrint16](https://user-images.githubusercontent.com/2170248/116131295-2ad2cd80-a707-11eb-8d0a-6f7912456e03.png)
+
+#### Allways call `matrix_output_unselect_delay()` (show `matrix_scan()` time, default MATRIX_IO_DELAY)
+ - `make MTEST=matrix_debug_scan,allways_delay handwired/symmetric70_proto/proton_c/normal:default:flash`
```
Comment (tzarc) [inline keyboards/handwired/symmetric70_proto/matrix_debug/readme.md:108]:
```suggestion
#### Always call `matrix_output_unselect_delay()` (show `matrix_scan()` time, default MATRIX_IO_DELAY)
 - `make MTEST=matrix_debug_scan,always_delay handwired/symmetric70_proto/proton_c/normal:default:flash`
```

File: keyboards/handwired/symmetric70_proto/matrix_fast/matrix_config_expand.c [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+#endif
+
+#include "cpp_map.h"
+
+#if defined(MATRIX_EXTENSION_74HC157) || defined(MATRIX_EXTENSION_74HC153)
+#   define  MATRIX_EXTANSION "matrix_extension_74hc15x.c"
+#endif
+
+#ifdef MATRIX_EXTANSION
+#    include MATRIX_EXTANSION
```
Comment (tzarc) [inline keyboards/handwired/symmetric70_proto/matrix_fast/matrix_config_expand.c:41]:
```suggestion
#   define  MATRIX_EXTENSION "matrix_extension_74hc15x.c"
#endif

#ifdef MATRIX_EXTENSION
#    include MATRIX_EXTENSION
```

File: keyboards/handwired/symmetric70_proto/proton_c/rules.mk [hunk 11]
```diff
@@ -0,0 +1,22 @@
+# MCU name
+MCU = STM32F303
+
+# Bootloader selection
+# BOOTLOADER =
```
Comment (fauxpark) [inline keyboards/handwired/symmetric70_proto/proton_c/rules.mk:5]:
```suggestion
BOOTLOADER = stm32-dfu
```

File: keyboards/handwired/symmetric70_proto/proton_c/rules.mk [hunk 12]
```diff
@@ -0,0 +1,22 @@
+# MCU name
+MCU = STM32F303
```
Comment (fauxpark) [inline keyboards/handwired/symmetric70_proto/proton_c/rules.mk:2]:
```suggestion
MCU = STM32F303
BOARD = QMK_PROTON_C
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by tzarc — CHANGES_REQUESTED** [review]
Fair few source files are also missing copyright headers, should add those too.

**Review by tzarc — APPROVED** [review]

**Pull request merged**